### PR TITLE
🎉 New Source: Visma e-conomic [python cdk]

### PR DESCRIPTION
**Pull request by sofiehellmark**
## What
Adding Visma e-conomic as a source. See issue https://github.com/airbytehq/connector-contest/issues/162.

## How
Using python cdk
Streams implemented:
 * [accounts](https://restdocs.e-conomic.com/#get-accounts)
* [customers](https://restdocs.e-conomic.com/#get-customers)
* [invoices booked](https://restdocs.e-conomic.com/#get-invoices-booked)
* [invoices booked document](https://restdocs.e-conomic.com/#get-invoices-booked-bookedinvoicenumber)
* [invoices paid](https://restdocs.e-conomic.com/#get-invoices-paid)
* [invoices total](https://restdocs.e-conomic.com/#get-invoices-totals)
* [products](https://restdocs.e-conomic.com/#get-products)

## Recommended reading order
1. `spec.yaml`
2. `schemas`

## 🚨 User Impact 🚨
N/A

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*
Screenshot 2022-10-30 at 17 20 59
</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

Screenshot 2022-10-31 at 17 50 32

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*
**Same as integration tests above**
Screenshot 2022-10-31 at 17 50 32

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks for your contribution @sofiehellmark ! Please request my review when the code is ready to review. If you have any questions feel free to ask them in our Slack channel #hacktoberfest-2022. You can get a link to access Slack here: https://airbyte.com/connector-contest

**Comment by marcosmarxm** [comment]
@sofiehellmark can you sign the CLA?

**Comment by sofiehellmark** [comment]
Hi! Ready for review on my first PR here 😄 I have updated the documentation now, let me know if I missed something. About the tests - could you please guide me if there is more I need to implement/provide?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Hello @sofiehellmark, Marcos from Airbyte here 👋 . We received more than 25 new contributions along the weekend. One is yours 🎉  thank so much for! Our team is limited and maybe the review process can take longer than expected. As described in the Airbyte's Hacktoberfest your contribution was submitted before November 2nd and it is eligible to win the prize. The review process will validate other requirements. I ask to you patience until someone from the team review it. 

Because I reviewed some contributions for Hacktoberfest so far I saw some common patterns you can check in advance:
* [ ] Make sure you have added connector documentation to `/docs/integrations/`
* [ ] Remove the file `catalog` from `/integration_tests`
* [ ] Edit the `sample_config.json` inside `/integration_tests`
* [ ] For the `configured_catalog` you can use only `json_schema: {}`
* [ ] Add title to all properties in the `spec.yaml`
* [ ] Make sure the `documentationUrl` in the `spec.yaml` redirect to Airbyte's future connector page, eg: connector Airtable the `documentationUrl: https://docs.airbyte.com/integrations/sources/airtable`
* [ ] Review now new line at EOF (end-of-file) for all files.

If possible send to me a DM in Slack with the tests credentials, this process will make easier to us run integration tests and publish your connector. If you only has production keys, make sure to create a bootstrap.md explaining how to get the keys.

File: airbyte-integrations/builds.md [hunk 1]
```diff
@@ -2,133 +2,134 @@
 
 # Sources
 
-|name |status |
-| :--- | :--- |
-| 3PL Central  | [![source-amazon-seller-partner](https://img.shields.io/endpoint?url=https%3A%2F%2Fdnsgjos7lj2fu.cloudfront.net%2Ftests%2Fsummary%2Fsource-tplcentral%2Fbadge.json)](https://dnsgjos7lj2fu.cloudfront.net/tests/summary/source-tplcentral) |
```
Comment (marcosmarxm) [inline airbyte-integrations/builds.md:7]: Can you check this file change?

File: airbyte-integrations/connectors/source-visma-economic/integration_tests/catalog.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
+  "streams": [
+    {
+      "name": "TODO fix this file",
+      "supported_sync_modes": ["full_refresh", "incremental"],
+      "source_defined_cursor": true,
+      "default_cursor_field": "column1",
+      "json_schema": {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "type": "object",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/integration_tests/catalog.json:10]: Remove this file.

File: airbyte-integrations/connectors/source-visma-economic/integration_tests/sample_config.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "fix-me": "TODO"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/integration_tests/sample_config.json:2]: Please update this file.

File: airbyte-integrations/connectors/source-visma-economic/requirements.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+-e .
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/requirements.txt:1]:
```
# This file is autogenerated -- only edit if you know what you are doing. Use setup.py for declaring dependencies.
-e ../../bases/source-acceptance-test
-e .
```

File: airbyte-integrations/connectors/source-visma-economic/source_visma_economic/sample_files/demo_config.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+"app_secret_token": "demo",
+"agreement_grant_token": "demo"
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/source_visma_economic/sample_files/demo_config.json:4]: Replace this with `sample_config.json`

File: airbyte-integrations/connectors/source-visma-economic/source_visma_economic/schemas/accounts.json [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  "properties": {
+    "accountNumber": {
+      "type": "integer",
+      "filterable": true,
+      "sortable": true,
+      "defaultsorting": "ascending",
+      "description": "The account's number."
+    },
+    "accountType": {
+      "enum": ["profitAndLoss", "status", "totalFrom", "heading", "headingStart", "sumInterval", "sumAlpha"],
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/source_visma_economic/schemas/accounts.json:16]: you must specify the type: string

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-visma-economic/source_visma_economic/schemas/accounts.json:16]: check all enums.

File: airbyte-integrations/connectors/source-visma-economic/source_visma_economic/source.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        self.agreement_grant_token: str = agreement_grant_token
+        super().__init__()
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        response_json = response.json()
+        if 'pagination' in response_json:
+            if 'nextPage' in response_json['pagination']:
+                parsed_url = urlparse(response_json['pagination']['nextPage'])
+                query_params = parse_qs(parsed_url.query)
+                return query_params
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/source_visma_economic/source.py:31]:
```suggestion
        if 'nextPage' in response_json.get('pagination').keys():
            parsed_url = urlparse(response_json['pagination']['nextPage'])
            query_params = parse_qs(parsed_url.query)
             return query_params
```

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-visma-economic/source_visma_economic/source.py:31]: You can simplify the code. Please add one unit test to make sure this is working.

File: airbyte-integrations/connectors/source-visma-economic/source_visma_economic/spec.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+documentationUrl: https://apis.e-conomic.com/
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/source_visma_economic/spec.yaml:1]: This is the Airbyte's connector page: `https://docs.airbyte.com/integrations/sources/visma-economic`

File: airbyte-integrations/connectors/source-visma-economic/source_visma_economic/spec.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+connectionSpecification:
+  $schema: http://json-schema.org/draft-07/schema#
+  title: Visma E-conomic Spec
+  type: object
+  required:
+    - app_secret_token
+    - agreement_grant_token
+  properties:
+    app_secret_token:
+      type: string
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-visma-economic/source_visma_economic/spec.yaml:11]: add title to all parameters.

File: docs/integrations/README.md [hunk 10]
```diff
@@ -178,6 +178,7 @@ For more information about the grading system, see [Product Release Stages](http
 | [Typeform](sources/typeform.md)                                                             | Alpha                | Yes                 |
 | [US Census](sources/us-census.md)                                                           | Alpha                | Yes                 |
 | [VictorOps](sources/victorops.md)                                                           | Alpha                | No                  |
+| [Visma e-conomic](sources/visma-economic.md)                                                | Alpha                | Yes                 |
```
Comment (marcosmarxm) [inline docs/integrations/README.md:181]:
```suggestion
| [Visma e-conomic](sources/visma-economic.md)                                                | Alpha                | No                 |
```

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments.

**Comment by sofiehellmark** [comment]
Hi @marcosmarxm, I have pushed requested changes. Thanks!

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
@sofiehellmark I read the Visma documentation (https://restdocs.e-conomic.com/#filtering) and found the API allows filtering. This allows to implement incremental syncs, which is much better to use the connector. What do you think about implementing this? I can help you whenever you need

**Comment by sofiehellmark** [comment]
Sounds good! I'll have a go at it, will let you know if any questions

**Comment by marcosmarxm** [comment]
Hello! I'm going to be out of the office this Friday and won't be able to review your contribution again today, I return next Monday. So far, most contributions look solid and are almost done to be approved. As said in [Chris' comment](https://airbytehq.slack.com/archives/C0450PCSDB2/p1667430678866629) all contributions made before 2-November are eligible to receive the prize and have 2 weeks to merge the contributions. But I ensure next week we're going to have your contribution merged. If you have questions about the implementation you can send them in `#hacktoberfest-2022` in Airbyte's Slack.

Sorry the inconvenience and see you again next week, thank you so much for your contribution!

**Review by marcosmarxm — APPROVED** [review]
Thanks @sofiehellmark

**Pull request merged**